src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,288 @@
+{
+  *outVector = Eigen::VectorXd::Zero(inVector.size());
+  for (auto it = indexMap.begin(); it != indexMap.end(); ++it)
+  {
+    (*outVector)(it->second) = inVector(it->first);
+  }
+}
+
+//=============================================================================
+dart::dynamics::Joint* findJointByName(
+    dart::dynamics::MetaSkeleton* metaSkeleton, const std::string& jointName)
+{
+  for (size_t i = 0; i < metaSkeleton->getNumJoints(); ++i)
+  {
+    if (metaSkeleton->getJoint(i)->getName() == jointName)
+      return metaSkeleton->getJoint(i);
+  }
+
+  return nullptr;
+}
+
+} // namespace
+
+//=============================================================================
+std::unique_ptr<SplineTrajectory> convertJointTrajectory(
+  const std::shared_ptr<MetaSkeletonStateSpace>& space,
+  const trajectory_msgs::JointTrajectory& jointTrajectory)
+{
+  if (!space)
+    throw std::invalid_argument{"StateSpace must be non-null."};
+
+  const auto numControlledJoints = space->getNumSubspaces();
+  if (jointTrajectory.joint_names.size() != numControlledJoints)
+  {
+    std::stringstream message;
+    message << "Incorrect number of joints: expected "
+        << numControlledJoints << ", got "
+        << jointTrajectory.joint_names.size() << ".";
+    throw std::invalid_argument{message.str()};
+  }
+
+  // Check that all joints are single DOF.
+  for (size_t i = 0; i < space->getNumSubspaces(); ++i)
+  {
+    auto joint = space->getJointSpace(i)->getJoint();
+    auto jointSpace = space->getSubspace(i);
+    auto rnJoint = dynamic_cast<RnJoint*>(jointSpace.get());
+    auto so2Joint = dynamic_cast<SO2Joint*>(jointSpace.get());
+
+    if (joint->getNumDofs() != 1 || (!rnJoint && !so2Joint))
+    {
+      std::stringstream message;
+      message << "Only single-DOF RnJoint and SO2Joint are supported. Joint "
+        << joint->getName() << "(index: " << i << ") is a "
+        << joint->getType() << " with "
+        << joint->getNumDofs() << " DOFs.";
+      throw std::invalid_argument{message.str()};
+    }
+  }
+
+  if (jointTrajectory.points.size() < 2)
+  {
+    throw std::invalid_argument{
+      "Trajectory must contain two or more waypoints."};
+  }
+
+  // Map joint indices between jointTrajectory and space subspaces.
+  std::map<size_t, size_t> rosJointToMetaSkeletonJoint;
+
+  auto metaSkeleton = space->getMetaSkeleton();
+
+  for (size_t i = 0; i < jointTrajectory.joint_names.size(); ++i)
+  {
+    std::string jointName = jointTrajectory.joint_names[i];
+    auto joint = findJointByName(metaSkeleton.get(), jointName);
+
+    if (!joint)
+    {
+      std::stringstream message;
+      message << "Joint " << jointName << " (index: " << i << ")"
+        << " does not exist in metaSkeleton.";
+      throw std::invalid_argument{message.str()};
+    }
+    auto index = metaSkeleton->getIndexOf(joint);
+    assert(index != dart::dynamics::INVALID_INDEX);
+
+    rosJointToMetaSkeletonJoint.emplace(std::make_pair(i, index));
```

Comment:
Verify that this is not a duplicate, i.e. that `rosJointToMetaSkeletonJoint[i]` does not already exist. You can do this by checking the `bool` flag in the `std::pair` returned by `emplace`.

Nit: Also, it would be nice to more descriptive variable names than `i` and `index`. 😉 Maybe `metaSkeletonIndex` and `trajectoryIndex`?